pkg/apiserver/compression.go

```diff
@@ -54,58 +55,78 @@ func configureCompression(router *gin.Engine, settings CompressionSettings) erro
 	}
 
 	router.Use(ginGzip.Gzip(level, opts...))
+	router.Use(recordResponseSize)
 
 	return nil
 }
 
 func parseLevel(level string) (int, error) {
 	switch level {
 	case "none":
-		return compressGzip.NoCompression, nil
+		return gzip.NoCompression, nil
 	case "default":
-		return compressGzip.DefaultCompression, nil
+		return gzip.DefaultCompression, nil
 	case "best":
-		return compressGzip.BestCompression, nil
+		return gzip.BestCompression, nil
 	case "fast":
-		return compressGzip.BestSpeed, nil
+		return gzip.BestSpeed, nil
 	default:
-		if i, err := strconv.ParseInt(level, 10, 64); err != nil {
+		if parsedLevel, err := strconv.ParseInt(level, 10, 64); err != nil {
 			return 0, fmt.Errorf("failed to parse level %s: %w", level, err)
+		} else if parsedLevel < gzip.NoCompression || parsedLevel > gzip.BestCompression {
+			return 0, fmt.Errorf("invalid compression level %d", parsedLevel)
 		} else {
-			return int(i), nil
+			return int(parsedLevel), nil
 		}
 	}
 }
 
-type gzipBodyReader struct {
-	body   io.ReadCloser
-	reader *compressGzip.Reader
-}
-
-func (r gzipBodyReader) Read(p []byte) (int, error) {
-	return r.reader.Read(p)
-}
+func decompressionFn(c *gin.Context) {
+	body := &countingBodyReader{
+		ReadCloser:          c.Request.Body,
+		readCompressedBytes: 0,
+	}
+	reader, err := gzip.NewReader(body)
 
-func (r gzipBodyReader) Close() error {
-	err := r.reader.Close()
 	if err != nil {
-		return err
+		// the body is not a proper gzip encoded body, so don't do anything
+		// the client most likely set the wrong content encoding on the message
+		return
```

Comment:
shouldn't we actually let it fail to adhere to standards? i would think that we are not supposed to "fix" something that is wrong from the client side

Reply:
Hmm, the problem here is for me how to handle this properly. I can't return an error because the library doesn't expect an error return. So I have 3 options:
- ignore the error as we do (and did in the past). the request will either fail once we try to parse the body as json or similar or it will succeed (if the content-encoding header was just bogus). kinda in line with specifying any other content-encoding header we currently don't handle
- create a special reader which immediately returns the correct error
- add it to the errors of the gin request. I guess we will print this at the end, but still continue to handle the request